Metabase can't decrypt the data when adding an encryption key after a failed migration

### Describe the bug

If you start a migration without an encryption key and you add it afterwards, it seems that Metabase isn't using it

### To Reproduce

1) Start v48.9 with an encryption key, initialize it
2) Remove encryption key and start 49.10, it won't work, as it doesn't have the encryption key
3) Now add the key and start 49.10 again, it won't work

### Expected behavior

It should work

### Logs

```

2024-05-13 21:46:53,652 ERROR liquibase.changelog :: ChangeSet migrations/001_update_migrations.yaml::v49.2024-04-09T10:00:03::qnkhuat encountered an exception.

UPDATE SUMMARY
Run:                          2
Previously run:             268
Filtered out:                17
-------------------------------
Total change sets:          287


FILTERED CHANGE SETS SUMMARY
DBMS mismatch:               17

2024-05-13 21:46:53,682 ERROR metabase.core :: Metabase Initialization FAILED
liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset migrations/001_update_migrations.yaml::v49.2024-04-09T10:00:03::qnkhuat:
     Reason: clojure.lang.ExceptionInfo: Unrecognized token 'KglemLU5XpyL': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 13] {:toucan2/context-trace [["resolve connection" {:toucan2.connection/connectable org.postgresql.jdbc.PgConnection}] ["resolve connection" {:toucan2.connection/connectable nil}]]}
	at liquibase.command.CommandScope.execute(CommandScope.java:253)
	at liquibase.Liquibase.lambda$update$0(Liquibase.java:245)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.Liquibase.runInScope(Liquibase.java:1419)
	at liquibase.Liquibase.update(Liquibase.java:234)
	at liquibase.Liquibase.update(Liquibase.java:212)
	at liquibase.Liquibase.update(Liquibase.java:194)
	at metabase.db.liquibase$migrate_up_if_needed_BANG_.invokeStatic(liquibase.clj:305)
	at metabase.db.liquibase$migrate_up_if_needed_BANG_.invoke(liquibase.clj:287)
	at metabase.db.setup$migrate_BANG_$fn__51146.invoke(setup.clj:80)
	at metabase.db.liquibase$do_with_liquibase$f_STAR___48822.invoke(liquibase.clj:139)
	at metabase.db.liquibase$do_with_liquibase.invokeStatic(liquibase.clj:142)
	at metabase.db.liquibase$do_with_liquibase.invoke(liquibase.clj:130)
	at metabase.db.setup$migrate_BANG_.invokeStatic(setup.clj:75)
	at metabase.db.setup$migrate_BANG_.doInvoke(setup.clj:56)
	at clojure.lang.RestFn.invoke(RestFn.java:445)
	at metabase.db.setup$run_schema_migrations_BANG_.invokeStatic(setup.clj:147)
	at metabase.db.setup$run_schema_migrations_BANG_.invoke(setup.clj:141)
	at metabase.db.setup$setup_db_BANG_$fn__51174$fn__51175.invoke(setup.clj:165)
	at metabase.util.jvm$do_with_us_locale.invokeStatic(jvm.clj:239)
	at metabase.util.jvm$do_with_us_locale.invoke(jvm.clj:225)
	at metabase.db.setup$setup_db_BANG_$fn__51174.invoke(setup.clj:160)
	at metabase.db.setup$setup_db_BANG_.invokeStatic(setup.clj:159)
	at metabase.db.setup$setup_db_BANG_.invoke(setup.clj:153)
	at metabase.db$setup_db_BANG_$fn__51194.invoke(db.clj:69)
	at metabase.db$setup_db_BANG_.invokeStatic(db.clj:64)
	at metabase.db$setup_db_BANG_.invoke(db.clj:55)
	at metabase.core$init_BANG__STAR_.invokeStatic(core.clj:116)
	at metabase.core$init_BANG__STAR_.invoke(core.clj:101)
	at metabase.core$init_BANG_.invokeStatic(core.clj:159)
	at metabase.core$init_BANG_.invoke(core.clj:154)
	at metabase.core$start_normally.invokeStatic(core.clj:171)
	at metabase.core$start_normally.invoke(core.clj:165)
	at metabase.core$entrypoint.invokeStatic(core.clj:204)
	at metabase.core$entrypoint.doInvoke(core.clj:198)
	at clojure.lang.RestFn.invoke(RestFn.java:397)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.Var.applyTo(Var.java:705)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.bootstrap$_main.invokeStatic(bootstrap.clj:31)
	at metabase.bootstrap$_main.doInvoke(bootstrap.clj:28)
	at clojure.lang.RestFn.invoke(RestFn.java:397)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at metabase.bootstrap.main(Unknown Source)
Caused by: liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset migrations/001_update_migrations.yaml::v49.2024-04-09T10:00:03::qnkhuat:
     Reason: clojure.lang.ExceptionInfo: Unrecognized token 'KglemLU5XpyL': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 13] {:toucan2/context-trace [["resolve connection" {:toucan2.connection/connectable org.postgresql.jdbc.PgConnection}] ["resolve connection" {:toucan2.connection/connectable nil}]]}
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:151)
	at liquibase.command.core.AbstractUpdateCommandStep.lambda$run$0(AbstractUpdateCommandStep.java:110)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.command.core.AbstractUpdateCommandStep.run(AbstractUpdateCommandStep.java:108)
	at liquibase.command.core.UpdateCommandStep.run(UpdateCommandStep.java:105)
	at liquibase.command.CommandScope.execute(CommandScope.java:217)
	... 49 more
Caused by: liquibase.exception.MigrationFailedException: Migration failed for changeset migrations/001_update_migrations.yaml::v49.2024-04-09T10:00:03::qnkhuat:
     Reason: clojure.lang.ExceptionInfo: Unrecognized token 'KglemLU5XpyL': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 13] {:toucan2/context-trace [["resolve connection" {:toucan2.connection/connectable org.postgresql.jdbc.PgConnection}] ["resolve connection" {:toucan2.connection/connectable nil}]]}
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:797)
	at liquibase.changelog.visitor.UpdateVisitor.executeAcceptedChange(UpdateVisitor.java:119)
	at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:68)
	at liquibase.changelog.ChangeLogIterator$2.lambda$run$0(ChangeLogIterator.java:133)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.changelog.ChangeLogIterator$2.run(ChangeLogIterator.java:122)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.Scope.child(Scope.java:252)
	at liquibase.Scope.child(Scope.java:256)
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:89)
	... 57 more
Caused by: clojure.lang.ExceptionInfo: Unrecognized token 'KglemLU5XpyL': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 13] {:toucan2/context-trace [["resolve connection" {:toucan2.connection/connectable org.postgresql.jdbc.PgConnection}] ["resolve connection" {:toucan2.connection/connectable nil}]]}
	at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:2481)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:762)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._reportInvalidToken(ReaderBasedJsonParser.java:3042)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._handleOddValue(ReaderBasedJsonParser.java:2085)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken(ReaderBasedJsonParser.java:812)
	at cheshire.parse$parse.invokeStatic(parse.clj:90)
	at cheshire.parse$parse.invoke(parse.clj:88)
	at cheshire.core$parse_string.invokeStatic(core.clj:208)
	at cheshire.core$parse_string.invoke(core.clj:194)
	at cheshire.core$parse_string.invokeStatic(core.clj:205)
	at cheshire.core$parse_string.invoke(core.clj:194)
	at metabase.db.custom_migrations.DeleteScanFieldValuesTriggerForDBThatTurnItOff$with_connection_STAR___48605$with_transaction_STAR___48606$fn__48607.invoke(custom_migrations.clj:1079)
	at clojure.core$filter$fn__5962.invoke(core.clj:2834)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5467.invokeStatic(core.clj:139)
	at clojure.core$seq__5467.invoke(core.clj:139)
	at metabase.db.custom_migrations.DeleteScanFieldValuesTriggerForDBThatTurnItOff$with_connection_STAR___48605$with_transaction_STAR___48606.invoke(custom_migrations.clj:1079)
	at toucan2.connection$bind_current_connectable_fn$fn__21143.invoke(connection.clj:104)
	at metabase.db.connection$do_transaction$thunk__32323.invoke(connection.clj:150)
	at metabase.db.connection$do_transaction.invokeStatic(connection.clj:165)
	at metabase.db.connection$do_transaction.invoke(connection.clj:146)
	at metabase.db.connection$do_with_transaction_primary_method_java_sql_Connection.invokeStatic(connection.clj:199)
	at metabase.db.connection$do_with_transaction_primary_method_java_sql_Connection.invoke(connection.clj:172)
	at clojure.lang.AFn.applyToHelper(AFn.java:165)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:457)
	at clojure.core$partial$fn__5908.invoke(core.clj:2643)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at methodical.impl.combo.threaded$fn__18233$fn__18234$fn__18237.invoke(threaded.clj:71)
	at methodical.impl.combo.threaded$reducer_fn$fn__18203$fn__18207.invoke(threaded.clj:23)
	at clojure.lang.ArrayChunk.reduce(ArrayChunk.java:58)
	at clojure.core.protocols$fn__8244.invokeStatic(protocols.clj:136)
	at clojure.core.protocols$fn__8244.invoke(protocols.clj:124)
	at clojure.core.protocols$fn__8204$G__8199__8213.invoke(protocols.clj:19)
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
	at clojure.core.protocols$fn__8236.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__8236.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6887)
	at clojure.core$reduce.invoke(core.clj:6869)
	at methodical.impl.combo.threaded$reducer_fn$fn__18203.invoke(threaded.clj:21)
	at clojure.core$comp$fn__5876.invoke(core.clj:2588)
	at methodical.impl.combo.threaded$combine_with_threader$fn__18213.invoke(threaded.clj:44)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invokeStatic(connection.clj:249)
	at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invoke(connection.clj:245)
	at clojure.lang.AFn.applyToHelper(AFn.java:165)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:457)
	at clojure.core$partial$fn__5908.invoke(core.clj:2643)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:58)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:195)
	at metabase.db.custom_migrations.DeleteScanFieldValuesTriggerForDBThatTurnItOff$with_connection_STAR___48605.invoke(custom_migrations.clj:1073)
	at toucan2.connection$bind_current_connectable_fn$fn__21143.invoke(connection.clj:104)
	at toucan2.connection$bind_current_connectable_fn$fn__21143.invoke(connection.clj:104)
	at toucan2.jdbc.connection$do_with_connection_primary_method_java_sql_Connection.invokeStatic(connection.clj:13)
	at toucan2.jdbc.connection$do_with_connection_primary_method_java_sql_Connection.invoke(connection.clj:11)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at clojure.core$partial$fn__5908.invoke(core.clj:2642)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at methodical.impl.combo.threaded$fn__18233$fn__18234$fn__18235.invoke(threaded.clj:70)
	at methodical.impl.combo.threaded$reducer_fn$fn__18203$fn__18207.invoke(threaded.clj:23)
	at clojure.lang.ArrayChunk.reduce(ArrayChunk.java:58)
	at clojure.core.protocols$fn__8244.invokeStatic(protocols.clj:136)
	at clojure.core.protocols$fn__8244.invoke(protocols.clj:124)
	at clojure.core.protocols$fn__8204$G__8199__8213.invoke(protocols.clj:19)
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
	at clojure.core.protocols$fn__8236.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__8236.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6887)
	at clojure.core$reduce.invoke(core.clj:6869)
	at methodical.impl.combo.threaded$reducer_fn$fn__18203.invoke(threaded.clj:21)
	at clojure.core$comp$fn__5876.invoke(core.clj:2587)
	at methodical.impl.combo.threaded$combine_with_threader$fn__18213.invoke(threaded.clj:43)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at clojure.core$partial$fn__5908.invoke(core.clj:2642)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:193)
	at toucan2.connection$do_with_connection_primary_method_.invokeStatic(connection.clj:204)
	at toucan2.connection$do_with_connection_primary_method_.invoke(connection.clj:194)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at clojure.core$partial$fn__5908.invoke(core.clj:2642)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at methodical.impl.combo.threaded$fn__18233$fn__18234$fn__18235.invoke(threaded.clj:70)
	at methodical.impl.combo.threaded$reducer_fn$fn__18203$fn__18207.invoke(threaded.clj:23)
	at clojure.lang.ArrayChunk.reduce(ArrayChunk.java:58)
	at clojure.core.protocols$fn__8244.invokeStatic(protocols.clj:136)
	at clojure.core.protocols$fn__8244.invoke(protocols.clj:124)
	at clojure.core.protocols$fn__8204$G__8199__8213.invoke(protocols.clj:19)
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
	at clojure.core.protocols$fn__8236.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__8236.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6887)
	at clojure.core$reduce.invoke(core.clj:6869)
	at methodical.impl.combo.threaded$reducer_fn$fn__18203.invoke(threaded.clj:21)
	at clojure.core$comp$fn__5876.invoke(core.clj:2587)
	at methodical.impl.combo.threaded$combine_with_threader$fn__18213.invoke(threaded.clj:43)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at clojure.core$partial$fn__5908.invoke(core.clj:2642)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:193)
	at metabase.db.custom_migrations.DeleteScanFieldValuesTriggerForDBThatTurnItOff.execute(custom_migrations.clj:1073)
	at liquibase.change.custom.CustomChangeWrapper.generateStatements(CustomChangeWrapper.java:169)
	at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1271)
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:755)
	... 72 more
Caused by: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'KglemLU5XpyL': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 13]
	... 234 more
2024-05-13 21:46:53,901 INFO metabase.core :: Metabase Shutting Down ...
```

### Information about your Metabase installation

```JSON
v48 migrating to v49
```


### Severity

P1

### Additional context

NA

--- a/src/metabase/db/custom_migrations.clj
+++ b/src/metabase/db/custom_migrations.clj
@@ -22,9 +22,9 @@
    [medley.core :as m]
    [metabase.config :as config]
    [metabase.db.connection :as mdb.connection]
-   [metabase.models.interface :as mi]
    [metabase.plugins.classloader :as classloader]
    [metabase.util.date-2 :as u.date]
+   [metabase.util.encryption :as encryption]
    [metabase.util.honey-sql-2 :as h2x]
    [metabase.util.log :as log]
    [toucan2.core :as t2]
@@ -104,6 +104,38 @@
   ^String [s]
   (when s
     (.toLowerCase (str s) Locale/US)))
+
+(defn json-in
+  "Default in function for columns given a Toucan type `:json`. Serializes object as JSON."
+  [obj]
+  (if (string? obj)
+    obj
+    (json/generate-string obj)))
+
+(defn- json-out [s keywordize-keys?]
+  (if (string? s)
+    (try
+      (json/parse-string s keywordize-keys?)
+      (catch Throwable e
+        (log/error e "Error parsing JSON")
+        s))
+    s))
+
+(def ^:private encrypted-json-in
+  (comp encryption/maybe-encrypt json-in))
+
+(defn- encrypted-json-out
+  [v]
+  (let [decrypted (encryption/maybe-decrypt v)]
+    (try
+      (json/parse-string decrypted true)
+      (catch Throwable e
+        (if (or (encryption/possibly-encrypted-string? decrypted)
+                (encryption/possibly-encrypted-bytes? decrypted))
+          (log/error e "Could not decrypt encrypted field! Have you forgot to set MB_ENCRYPTION_SECRET_KEY?")
+          (log/error e "Error parsing JSON"))  ; same message as in `json-out`
+        v))))
+
 
 ;;; +----------------------------------------------------------------------------------------------------------------+
 ;;; |                                                  MIGRATIONS                                                    |
@@ -716,9 +748,9 @@
 
 (define-reversible-migration MigrateDatabaseOptionsToSettings
   (let [update-one! (fn [{:keys [id settings options]}]
-                      (let [settings     (mi/encrypted-json-out settings)
-                            options      (mi/json-out-with-keywordization options)
-                            new-settings (mi/encrypted-json-in (merge settings options))]
+                      (let [settings     (encrypted-json-out settings)
+                            options      (json-out options true)
+                            new-settings (encrypted-json-in (merge settings options))]
                         (t2/query {:update :metabase_database
                                    :set    {:settings new-settings}
                                    :where  [:= :id id]})))]
@@ -729,12 +761,12 @@
                                                     [:not= :options "{}"]
                                                     [:not= :options nil]]})))
   (let [rollback-one! (fn [{:keys [id settings options]}]
-                        (let [settings (mi/encrypted-json-out settings)
-                              options  (mi/json-out-with-keywordization options)]
+                        (let [settings (encrypted-json-out settings)
+                              options  (json-out options true)]
                           (when (some? (:persist-models-enabled settings))
                             (t2/query {:update :metabase_database
                                        :set    {:options (json/generate-string (select-keys settings [:persist-models-enabled]))
-                                                :settings (mi/encrypted-json-in (dissoc settings :persist-models-enabled))}
+                                                :settings (encrypted-json-in (dissoc settings :persist-models-enabled))}
                                        :where  [:= :id id]}))))]
     (run! rollback-one! (t2/reducible-query {:select [:id :settings :options]
                                              :from   [:metabase_database]}))))
@@ -1080,7 +1112,7 @@
   ;; this migraiton to remove triggers for any existing DB that have this option on.
   ;; See #40715
   (when-let [;; find all dbs which are configured not to scan field values
-             dbs (seq (filter #(and (-> % :details (json/parse-string keyword) :let-user-control-scheduling)
+             dbs (seq (filter #(and (-> % :details encrypted-json-out :let-user-control-scheduling)
                                     (false? (:is_full_sync %)))
                               (t2/select :metabase_database)))]
     (classloader/the-classloader)

--- a/src/metabase/db/custom_migrations.clj
+++ b/src/metabase/db/custom_migrations.clj
@@ -122,9 +122,11 @@
     s))
 
 (def ^:private encrypted-json-in
+  "Should mirror [[metabase.models.interface/encrypted-json-in]]"
   (comp encryption/maybe-encrypt json-in))
 
 (defn- encrypted-json-out
+  "Should mirror [[metabase.models.interface/encrypted-json-out]]"
   [v]
   (let [decrypted (encryption/maybe-decrypt v)]
     (try
@@ -135,7 +137,6 @@
           (log/error e "Could not decrypt encrypted field! Have you forgot to set MB_ENCRYPTION_SECRET_KEY?")
           (log/error e "Error parsing JSON"))  ; same message as in `json-out`
         v))))
-
 
 ;;; +----------------------------------------------------------------------------------------------------------------+
 ;;; |                                                  MIGRATIONS                                                    |

--- a/test/metabase/db/custom_migrations_test.clj
+++ b/test/metabase/db/custom_migrations_test.clj
@@ -13,7 +13,6 @@
    [clojurewerkz.quartzite.triggers :as triggers]
    [medley.core :as m]
    [metabase.api.database-test :as api.database-test]
-   [metabase.config :as config]
    [metabase.db :as mdb]
    [metabase.db.custom-migrations :as custom-migrations]
    [metabase.db.schema-migrations-test.impl :as impl]
@@ -1706,6 +1705,6 @@
           (migrate!)
           (is (true? (encryption/possibly-encrypted-string? (db-detail)))))
 
-        (migrate! :down (dec (config/current-major-version)))
+        (migrate! :down 48)
         (testing "after migrate down, db details should still be encrypted"
           (is (true? (encryption/possibly-encrypted-string? (db-detail)))))))))

--- a/test/metabase/db/custom_migrations_test.clj
+++ b/test/metabase/db/custom_migrations_test.clj
@@ -1683,7 +1683,7 @@
                         (is (nil? (:cache_field_values_schedule db))))))))]
         (testing "without encryption key"
           (do-test))
-        (testing "without encryption key"
+        (testing "with encryption key"
           (encryption-test/with-secret-key "dont-tell-anyone-about-this"
             (do-test)))))))
 

--- a/test/metabase/db/custom_migrations_test.clj
+++ b/test/metabase/db/custom_migrations_test.clj
@@ -13,6 +13,7 @@
    [clojurewerkz.quartzite.triggers :as triggers]
    [medley.core :as m]
    [metabase.api.database-test :as api.database-test]
+   [metabase.config :as config]
    [metabase.db :as mdb]
    [metabase.db.custom-migrations :as custom-migrations]
    [metabase.db.schema-migrations-test.impl :as impl]
@@ -1637,46 +1638,74 @@
 (deftest delete-scan-field-values-trigger-test
   (testing "We should delete the triggers for DBs that are configured not to scan their field values\n"
     (impl/test-migrations "v49.2024-04-09T10:00:03" [migrate!]
-      (api.database-test/with-db-scheduler-setup
-        (let [db-with-full-schedules (new-instance-with-default :metabase_database
-                                                                {:metadata_sync_schedule      "0 0 * * * ? *"
-                                                                 :cache_field_values_schedule "0 0 1 * * ? *"
-                                                                 :is_full_sync                true
-                                                                 :is_on_demand                false})
-              db-manual-schedule     (new-instance-with-default :metabase_database
-                                                                {:details                     (json/generate-string {:let-user-control-scheduling true})
-                                                                 :is_full_sync                true
-                                                                 :is_on_demand                false
-                                                                 :metadata_sync_schedule      "0 0 * * * ? *"
-                                                                 :cache_field_values_schedule "0 0 2 * * ? *"})
-              db-on-demand           (new-instance-with-default :metabase_database
-                                                                {:details                     (json/generate-string {:let-user-control-scheduling true})
-                                                                 :is_full_sync                false
-                                                                 :is_on_demand                true
-                                                                 :metadata_sync_schedule      "0 0 * * * ? *"
-                                                                 :cache_field_values_schedule "0 0 2 * * ? *"})
-              db-never-scan          (new-instance-with-default :metabase_database
-                                                                {:details                     (json/generate-string {:let-user-control-scheduling true})
-                                                                 :is_full_sync                false
-                                                                 :is_on_demand                false
-                                                                 :metadata_sync_schedule      "0 0 * * * ? *"
-                                                                 :cache_field_values_schedule "0 0 2 * * ? *"})
-              db-with-scan-fv        [db-with-full-schedules db-manual-schedule]
-              db-without-scan-fv     [db-on-demand db-never-scan]]
-          (doseq [db (concat db-with-scan-fv db-without-scan-fv)]
-            (#'database/check-and-schedule-tasks-for-db! (t2/instance :model/Database db))
-            (testing "sanity check that the schedule exists"
-              (is (= (#'task.sync-databases-test/all-db-sync-triggers-name db)
-                     (#'task.sync-databases-test/query-all-db-sync-triggers-name db)))))
+      (letfn [(do-test []
+                (api.database-test/with-db-scheduler-setup
+                  (let [db-with-full-schedules (new-instance-with-default :metabase_database
+                                                                          {:metadata_sync_schedule      "0 0 * * * ? *"
+                                                                           :cache_field_values_schedule "0 0 1 * * ? *"
+                                                                           :is_full_sync                true
+                                                                           :is_on_demand                false})
+                        db-manual-schedule     (new-instance-with-default :metabase_database
+                                                                          {:details                     (json/generate-string {:let-user-control-scheduling true})
+                                                                           :is_full_sync                true
+                                                                           :is_on_demand                false
+                                                                           :metadata_sync_schedule      "0 0 * * * ? *"
+                                                                           :cache_field_values_schedule "0 0 2 * * ? *"})
+                        db-on-demand           (new-instance-with-default :metabase_database
+                                                                          {:details                     (json/generate-string {:let-user-control-scheduling true})
+                                                                           :is_full_sync                false
+                                                                           :is_on_demand                true
+                                                                           :metadata_sync_schedule      "0 0 * * * ? *"
+                                                                           :cache_field_values_schedule "0 0 2 * * ? *"})
+                        db-never-scan          (new-instance-with-default :metabase_database
+                                                                          {:details                     (json/generate-string {:let-user-control-scheduling true})
+                                                                           :is_full_sync                false
+                                                                           :is_on_demand                false
+                                                                           :metadata_sync_schedule      "0 0 * * * ? *"
+                                                                           :cache_field_values_schedule "0 0 2 * * ? *"})
+                        db-with-scan-fv        [db-with-full-schedules db-manual-schedule]
+                        db-without-scan-fv     [db-on-demand db-never-scan]]
+                    (doseq [db (concat db-with-scan-fv db-without-scan-fv)]
+                      (#'database/check-and-schedule-tasks-for-db! (t2/instance :model/Database db))
+                      (testing "sanity check that the schedule exists"
+                        (is (= (#'task.sync-databases-test/all-db-sync-triggers-name db)
+                               (#'task.sync-databases-test/query-all-db-sync-triggers-name db)))))
 
+                    (migrate!)
+                    (testing "default options and scan with manual schedules should have scan field values"
+                      (doseq [db db-with-scan-fv]
+                        (is (= (#'task.sync-databases-test/all-db-sync-triggers-name db)
+                               (#'task.sync-databases-test/query-all-db-sync-triggers-name db)))))
+
+                    (testing "never scan and on demand should not have scan field values"
+                      (doseq [db (t2/select :model/Database :id [:in (map :id db-without-scan-fv)])]
+                        (is (= #{(#'api.database-test/sync-and-analyze-trigger-name db)}
+                               (#'task.sync-databases-test/query-all-db-sync-triggers-name db)))
+                        (is (nil? (:cache_field_values_schedule db))))))))]
+        (testing "without encryption key"
+          (do-test))
+        (testing "without encryption key"
+          (encryption-test/with-secret-key "dont-tell-anyone-about-this"
+            (do-test)))))))
+
+(deftest migration-works-when-have-encryption-key-test
+  ;; this test is here to warn developers that they should test their migrations with and without encryption key
+  (encryption-test/with-secret-key "dont-tell-anyone-about-this"
+    ;; run migration to the latest migration
+    (impl/test-migrations ["v49.2024-04-09T10:00:03"] [migrate!]
+      ;; create a db because db.details should be encrypted
+      (let [db-id     (:id (new-instance-with-default :metabase_database {:details (encryption/maybe-encrypt "{}")}))
+            db-detail (fn []
+                        (:details (t2/query-one {:select [:details]
+                                                 :from   [:metabase_database]
+                                                 :where  [:= :id db-id]})))]
+        (testing "sanity check that db details is encrypted"
+          (is (true? (encryption/possibly-encrypted-string? (db-detail)))))
+
+        (testing "after migrate up, db details should still be encrypted"
           (migrate!)
-          (testing "default options and scan with manual schedules should have scan field values"
-            (doseq [db db-with-scan-fv]
-              (is (= (#'task.sync-databases-test/all-db-sync-triggers-name db)
-                     (#'task.sync-databases-test/query-all-db-sync-triggers-name db)))))
+          (is (true? (encryption/possibly-encrypted-string? (db-detail)))))
 
-          (testing "never scan and on demand should not have scan field values"
-            (doseq [db (t2/select :model/Database :id [:in (map :id db-without-scan-fv)])]
-              (is (= #{(#'api.database-test/sync-and-analyze-trigger-name db)}
-                     (#'task.sync-databases-test/query-all-db-sync-triggers-name db)))
-              (is (nil? (:cache_field_values_schedule db))))))))))
+        (migrate! :down (dec (config/current-major-version)))
+        (testing "after migrate down, db details should still be encrypted"
+          (is (true? (encryption/possibly-encrypted-string? (db-detail)))))))))

